contracts/CDPLeverage.sol

```diff
@@ -0,0 +1,456 @@
+
+        investors[msg.sender] = sender;
+
+        emit OpenPosition(msg.sender, msg.value, sender.purchPrice, _layers);
+    }
+
+
+    /**
+    * @dev Liquidate - Liqduiates the sender's position and returns ETH
+    * Possible outcomes:
+    * 1. USD/ETH drops below sender's price floor
+    * 2. USD/ETH appreciates relative to sender's purchase price
+    * 3. USD/ETH deppreciates relative to sender's purchase price
+    * Sender must send ETH with function call to cover the governance and stability fee with use of the Maker platform
+    * Any excess ETH will be returned with the closing of the position
+    */
+    function liquidate()
+        payable
+        onlyInvestor
+        public
+    {
+        uint256 currPriceEth = uint256(pip.read());                          // DAI/WETH In WAD
+
+        Investor memory sender = investors[msg.sender];
+
+        uint256 releasedPeth;
+        uint256 remainingPeth = tub.ink(sender.cdpID);                       // retrieve value of collatoral in CDP
+        uint256 payout;
+        uint256 wethAmount;
+        uint256 daiAmount = sender.daiAmountFinal;
+
+        // checks if CDP debt is 0
+        if (tub.tab(sender.cdpID) == 0) {
+            //*** CDP is auto-liquidated
+            //*** Convert to WETH and send back to investor = 1 + 2
+            //*** 1. Remaining PETH in debt-free CDP
+            //*** 2. Outstanding DAI from final layer
+
+            releaseWeth(sender.cdpID, remainingPeth);
+            wethAmount = marketBuy(weth, dai, sender.daiAmountFinal);
+            payout = add(remainingPeth,wethAmount);
+
+        } else {
+
+            //*** USD/ETH price deppreciated OR appreciated
+
+            require(0 < msg.value, "Ether is required to liqudiate.");
+
+            uint256 remainingDebt = sender.totalDebt;
+            uint256 remainingDai;
+            uint256 excessWeth;
+
+            IWETH(weth).deposit.value(msg.value)();
+
+            uint256 mkrFee;
+            uint256 wethFee;
+            (mkrFee, wethFee) = wipeFee(sender.cdpID, remainingDebt);
+
+            // Pay for governance fee
+            require(wethFee <= msg.value, "Not enough ether provided for fees");
+            dex.buyAllAmount(gov, mkrFee, weth, wethFee);
+            excessWeth = sub(msg.value,wethFee);
+
+            // wipe last layer of cdp
+            remainingDebt = wipeDebt(sender.cdpID, daiAmount, remainingDebt);
+            releasedPeth = wdiv(wmul(daiAmount,makerLR),sender.purchPrice);
+            releaseWeth(sender.cdpID, releasedPeth);
+            remainingPeth -= releasedPeth;
+
+            // empty remaining debt or reach dust pan condition (CDP peth = 0.0051), whichever comes first
+            while (remainingDebt > 0) {
+                daiAmount = marketBuy(dai, weth, releasedPeth);
+
+                if (daiAmount > remainingDebt) {
+                    remainingDai = sub(daiAmount,remainingDebt);
```

Comment:
Could we rename `remainingDai` to `excessDai` – this part tripped me up for a minute